### ISSUE TITLE
PDE remove J K JJ JK ...

### ISSUE THREAD

**Issue by TTitscher** [nutofem/nuto#164]
Disclaimer: This is the idea of @Psirus, so I cannot be sure to accurately represent his views. But he was ~~~too lazy~~~ not motivated enough to write it himself.

## Problem

`GlobalDofVector` and `GlobalDofSparseMatrix` are the main global types returned from the assembler. They are designed to meet the requirements for our [linear constraints](http://nuto.readthedocs.io/en/master/Constraints.html) that eliminate dofs from the global system. They do that by splitting the types in _independent_(`J`) and _dependent_(`K`) parts. For other types of constraints (penalty, lagrange multiplier, ...) this is not needed. 

## Solution

We make the subdivision into `J/K` not the default case, but a special case. The assembler returns the existing `DofVector` and `DofSparseMatrix` instead of `GlobalDofVector` and `GlobalSparseMatrix`. The latter ones are removed.
Possible use cases:
- penalty method:
    - chose an arbitrary dof numbering
    - assemble the system
    - apply penalty terms to some of those entries
- linear constraints:
    - chose a dof numbering that sorts dependent dofs to the end
    - assemble the system
    - do the calculations including `CMat` on `BlockViews` of the matrix, e.g. [`topLeftCorner`](https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#a5b8664897a39a39fbb8e98d70b94f16d) will correspond to `GlobalSparseMatrix.JJ`

In NuTo_Master those details are hidden in a `ApplyCMatrix` method. This method can be rewritten to work on those views instead of submatrices `JJ, KJ,...` and will be hidden.

## Discussion

Upon external request, I [benchmarked](https://gist.github.com/TTitscher/c122fb60b12c2ba0fa4e5e488a766003#file-applycmat-cpp) it. 

~~~
----------------------------------------------------------
Benchmark                   Time           CPU Iterations
----------------------------------------------------------
ApplyCMatFull       103761570 ns  103737108 ns          6
ApplyCMatSeparated  100617959 ns  100616271 ns          6
~~~
(For 2M dofs, 100 dependent, sparsity 0.0001. `Separated` means with JJ, JK, ...)

Even though it is a bit slower, IMO a solver call will dominate the run time by far.

**Comment by TTitscher** [comment]
I just figured out that this may not be as trivial as I thought - still doable however. If we regard multiple dof types `D` and `T`, we currently have:

~~~
DDJJ DTJJ | DDJK DTJK 
TDJJ TTJJ | TDJK TTJK 
----------+----------
DDKJ DTKJ | DDKK DTKK 
TDKJ TTKJ | TDKK TTKK 
~~~

In the non-separated idea, we have ( if we keep the dof numbering )
~~~
DD DT
TD TT 
~~~

Extracting `JJ` from that then actually requires 4 blocks, or 9 blocks if we have 3 dof types. Still doable, but not as straight forward as I originally thought.

**Comment by joergfunger** [comment]
The main reason for our implementation of splitting the nodal dof vector into j and k components was the computation of the modified Hessian/residual for linear constraints with multiple dofs and the computation of the trial state for problems with localization. Both problems could imo be solved differently, thus we only need to use a single global (block) vector/matrix.

The trial force is the force that would be obtained, if the constrained displacements where modified and the external force (the one that is only time dependent) is changed. This is complicated by the fact that  new velocities/accelerations also depend on this new displacement. In particular, the change in v/a is nonzero for zero d, thus all dofs (and not only the constrained ones) had to be considered.

New idea for computing the trial state in Newmark type problems
1. calculate the new residual for F(t+dt) (essentially the change in the external force since F(t)=0, but this also includes changes in e.g. creep where dt is an input) 
1. initialize the delta of the total dof vector with zero (dd=0)
1. loop over the constraint equations and modify the constrained ones
1. compute from dd the increments of dv and da of velocities and accelerations
1. compute the full hessians (0,1,2)
1. compute the total trial force F_trial = F(t+dt) + H0*dd + H1*dv + H2*da
1. compute the full Hessian (all dofs) by summing up the contributions from 0,1,2 according to the time integrations scheme 
1. Solve (dd = Solve(H_full, F_full, LinConstraints))

Remarks
- in 1., we essentially compute dF/dt with finite differences, this could also be replaced by requiring a function dF/dt directly (but I don't think this is worth the effort)
- the numbering of the dofs should not depend on the constraints. Thus, any solution vector (all dofs) is identical , even if different boundary conditions are applied
- I would store all nodes of a specific interpolation at this dof type (in the class). This essentially describes the function space we are defining for a specific dof type
- the node dofs are numbered when they are inserted, starting from zero (for each subdomain). We do not even have to number each dof separately, but if all nodes in this dof type have a constant number dofs (say 3 displacements) then the node number should be sufficient to compute the dofs.
- any instance of the solution vector (this could also be v and a since they live in the same function space) knows its functions space (and vise versa). Thus, if we have to move some nodes to a different domain (e.g. due to rebalancing of subdomains) we could do so via this dof-class
- constraints also refer to this function space
- there is a container class that stores all constraints. This class provides a method that has as input the full vector/matrix, and provides as output the reduced contributions (F_mod, H_mod)
- the computation of the reduction could be done by adding the unit matrix (dim jj) to C, and then just always compute Hmod = C^T H C or F_mod = C^T F. We should check, how time consuming this is. Alternatively, we will have to assemble the domain block matrix into a global matrix appropriate for the solver anyway (at least when we have multiple dofs). For trilinos, this is also done on an entry by entry base, thus those single values could be multiplied by the constraints (only required if one of the rows/cols of this matrix entry is constrained)
- the numbering for the reduced solution space is only internal to this constraint container (and thus the ordering in the constraint matrix C) and can be either recomputed each time a solve is called or stored as a member there.
- we could even think of getting rid of the block vectors/matrices, but this is maybe another issue :)

**Comment by TTitscher** [comment]
The question for me is what _exactly_ we want to accomplish. 

- **Avoid J/K split** for more cosmetic reasons, like less classes, no more exports/imports and such. I understand that. 
I guess that the main reason for the J/K split was an _efficient_ calculation of the modified hessian/residual that, in most cases `cmat.numEntries == 0`, involves just picking JJ/J without any calculations. I like that and this will easily beat any other solution performance-wise. But I also think that your algorithm sketch will work. Benchmarks (~~~I'll do them soon~~~ _well, no idea how to do the CMat stuff_) will show the performance impact. If it is not _too bad_ and we decide to do that, the implementation should be straight forward.
By the way: This does _not_ require a constant dof numbering. And by using a clever dof numbering (dependent on the constraints) and `Eigen::block`, we might elimit J/K without major performance issues.

- **Keeping a rock solid numbering**. The reason here might be [Disclaimer: _The reasoning is not totally clear to me._] a GlobalDofVector that does not change its numbering. Reasons to change the numbering may be:
    1) _changing constraints_: I _can_ think of use cases for that, but only very few. Anyways, this can be solved by removing J/K and the algorithm you pointed out. 
    2) _MPI load balancing_: This actually changes the number of dofs in a domain. Not possible to keep a constant numbering.
    3) _local mesh refinement_: Also changes the number of dofs as above.

First, I want to mention that no one actually implemented 2) and 3), not even on `OldNuTo`. So we have no idea how that will work out. Anyways, for 2) you presented a combination of a `GlobalDofValue` vector (_solution vector is misleading IMO_) and a vector space and constraints that handle all the renumbering/moving required. But then:
- If you say that we can somehow manage that renumbering, why not also include the renumbering based on constraint changes? I expect this to be much easier compared to a load balancing renumbering.
- Moving elements/dofs/nodes and their renumbering is certainly not enough. More _magic_ is required anyways, like moving/reallocating history data or cells.

So I am certainly in for the J/K removal. This does simplify writing time integration schemes because constraints can be handled at exactly one place outside of them. And I do not think that changing the constraints should be _everyday feature_. It should be possible, sure, but if that involves some internal renumbering (of nodes and possibly the `GlobalDofValue` vector), I am fine with that.

**Comment by joergfunger** [comment]
The reason for the presented **rock solid numbering** and the **removal of the j/k components** is primarily to separate the numbering used in the function evaluations (gradient, hessian) from the constraints. Thus, a vector of GlobalDofValues is always identical, no matter what constraints are applied.  In oldNuto, we performed after having created all constraints the numbering. The transfer of the gobalDofVector (say the initial conditions) was performed using the values stored at the nodes.

In new NuTo, I could image something like that
~~~cpp
//this function space essentially describes the approximation of certain variables (disp, T), the discretization and how to compute the approximation. 
class FunctionSpace 
{
     FunctionSpace(MeshFEM);
private:
     std::map<DofType,std::vector<Node*> > mDofNumbers;
};

// this node just describes the entries in the domain dof vector
template<int TDim>
class Node
{
     std::array<TDim> mDofNumberInDomainVector; //this is the entry for a dof in the domain(global) solution vector
};

// this is a discrete representation of the functionSpace, so e.g. the current displacement or temperature  state (or both)
// it includes not only the active, but all dofs
// no need for a block vector here
class GlobalDofVector
{
    GlobalDofVector(const FunctionSpace& functionSpace);
private:
    const FunctionSpace& mFunctionSpace;
    EigenVectorXd mDofValues; //
};

//boundary conditions
class BoundaryConditions
{
   BoundaryConditions(const FunctionSpace& functionSpace)
private:
   const FunctionSpace& mFunctionSpace;
   Constraints mConstraints;  //this is a list of constraint equations, each referring to the numbering in the 
in FunctionSpace
}
~~~

In the main file, someone could do
~~~cpp
//this class provides methods to call the gradient and the hessian on this function space
DofType DispDof("Displacement", 3);
DofType NonlocalStrainDof("NonlocalEqStrain", 1);
//create function space with both dof types
FunctionSpace myFunctionSpace (mesh, ...);
//create problem that solves only for the displacement variables (this could also be a list of dofs)
StaticProblemDisp myproblem(FunctionSpace myFunctionSpace, DispDof); 
//create problem that solves only for the nonlocal straints 
StaticProblemNonlocalStrain myproblem(FunctionSpace myFunctionSpace,NonlocalStrainDof); 

//create set of boundary conditions for nonlocal strains and displacements
boundaryConditionsNonlocalStrain 
..
boundaryConditionsDisp 
..
//create a discrete representation of the FunctionSpace 
GlobalDofVector dofValues(myFunctionSpace);

//set the initial conditions in dofValues
..

for (int timeStep=0; timeSteps<10;  timeStep++)
{
     dofValues = StaticSolver.DoStep(StaticProblemNonlocalStrain, dofValues, dt*timeStept, boundaryConditionsNonlocalStrain);
     dofValues = StaticSolver.DoStep(StaticProblemDisp, dofValues, dt*timeStep, boundaryConditionsDisp);
}
~~~
We could also "hide" these dofValues in the problem definition (dofValues is a member of the Problem). Thus, for the standard case of a single Solver this would look like
~~~cpp
//initial conditions
StaticProblemDisp.SetValues(dofValues);
for (int timeStep=0; timeSteps<10;  timeStep++)
{
     StaticSolver.DoStep(StaticProblemDisp, dt*timeStep, boundaryConditionsDisp);
}
~~~

In the case of coupling the nonlocal damage model with a dynamic Newmark scheme (second order), this would look like this
~~~cpp
//create problem that solves only for the displacement variables and its time derivatives
NewmarkProblemDisp myproblem(FunctionSpace myFunctionSpace, DispDof); 
//create problem that solves only for the nonlocal strains 
StaticProblemNonlocalStrain myproblem(FunctionSpace myFunctionSpace, NonlocalStrainDof); 

//create set of boundary conditions for nonlocal strains and displacements
boundaryConditionsNonlocalStrain 
..
boundaryConditionsDisp 
..
//create a discrete representation of the FunctionSpace 
GlobalDofVector dofValues(myFunctionSpace);
GlobalDofVector dofVelocities(myFunctionSpace);
GlobalDofVector dofAccelerations(myFunctionSpace);

//set the initial conditions in dofValues
..

for (int timeStep=0; timeSteps<10;  timeStep++)
{
     StaticSolver.DoStep(StaticProblemNonlocalStrain, dt*timeStept, boundaryConditionsNonlocalStrain);
     NewmarkProblemDisp.SetValues(StaticProblemNonlocalStrain.GetValues());
     NewmarkSolver.DoStep(NewmarkProblemDisp, dt*timeStep, boundaryConditionsDisp);
     StaticProblemNonlocalStrain.SetValues(NewmarkProblemDisp.GetValues());
}
~~~
The transfer of nodalDofValues is just a copy, since the numbering in GlobalDofVector does not depend on the constraints. If this depends on the numbering, I do not know how to transfer the nodal values. In a similar way, it is possible to have to consecutive sets of constrain conditions say first a shear test, and then a normal tension with the shear displacements being constant
~~~cpp
for (int timeStep=0; timeSteps<10;  timeStep++)
{
     StaticSolver.DoStep(StaticProblemDisp, dt*timeStep, boundaryConditionsDisp1);
}
for (int timeStep=10; timeSteps<21;  timeStep++)
{
     StaticSolver.DoStep(StaticProblemDisp, dt*timeStep, boundaryConditionsDisp2);
}
~~~
In the case of the numbering depending on the constraints, I don't know how to transfer the information between solution phase 1 and 2.

**Comment by TTitscher** [comment]
I find it hard think about that and this topic should be discussed in person IMO. And this seems like _faaar_ more than just removing J/K. Well, my thoughts:

For me, `Mesh` sounds like the _full_ function space. By applying constraints, I reduce the function space. (And I could convert function spaces into each other). And each function space is able to generate a global dof vector. We are then able to calculate the gradient from this dof vector. This would allow us to have, say, different function spaces (different constraints) and calculate gradients independently. 

But then, the question of history data pops up. Both function spaces work on the same history data, that are not passed as an argument. For me, this makes the idea of having _multiple, exchangable_ function spaces pointless. So can we add the constraints to the mesh and be done? And we obviously allow changing the constraints.

**Comment by joergfunger** [comment]
In my opinion, the Removal of jk components in the gobal matrices has he following advantages and disadvantages
_Advantages_
- There  won’t be any need to sort the solution vectors according to dependent/independent parts, we always have the complete dof state.
- This allows to remove the BlockVector/Blockmatrix on the global domain. As a consequence, the computations are simpler (we can use directly the linear algebra from eigen or any other matrix type on the global scale), the code is better to understand (less data types, no own implementation of linear algebra methods) 
- The constraints can be easily separated into the solution step and are only passed as a parameter to the solution procedure. Even though this might seem to be only a minor issue, this is very fundamental.
_Disadvantages_
- There is an additional mapping step from the initial (and constant) dof numbering to a constraint dependent numbering
- The extraction of the jk components from the full matrix is computationally more expensive (e.g. when computing the trial state)
```cpp
Newmark::DoStep(..)
{
   ComputeTrialSolution(t,BC);
...
}
Newmark::TrialSolution(t,BoundaryConditions)
{ 
   //Eigen::Vector can be replaced by a general representation of the global function space
   Eigen::Vector fullGradient = myNewmarkProblem.Gradient(prevDofSolution ,prevDofVelocity, prevDofAcceleration, prevHistSolution , t,t-prevt, dofs)
   Eigen::SparseMatrix fullHessian012 = myNewmarkProblem.Hessian0123(prevDofSolution ,prevDofVelocity, prevDofAcceleration, prevHistSolution , t,t-prevt, dofs) //separate routines
   //compute trial residual for deltaDof=0
   EigenVector trialResidual = fullGradient + fullHessian0*deltaDof + fullHessian1*deltaV + ..; 
   //Reduced System;
   C = BoundaryConditions.GetCMatrix(dof);
   hmod = C.transpose() * H * C //C = matrix of dimension (j, j+k);
   fmod = C.transpose() * trialResidual;
   solutionMod = Solver.Solve(hmod,hmod);
   solutionFull = C * solutionMod; 
   Loop over constraint equations and add bRHS for each constraint dof;
}
`

**Comment by joergfunger** [comment]
Implemented with PR #238
